### PR TITLE
Repo structure

### PR DESCRIPTION
**Pull request by hannahker**
Here is a suggested structure for organizing our code. Everything should be set up so that we contribute to either `r/R/` or `python/src/`, and ideally add in examples of how these utility functions should be used in the appropriate `examples/` directory. The top level `python/` and `r/` directories should be totally independent, but for now I like the idea of just having one place to go for all of this work.

### CODE REVIEW

**Comment by zackarno** [comment]
maybe under python 

- include a dummy `.env` file? 
- should we have the `requirements-dev.txt` file pre-built with black, isort, python-dotenv, jupyter-black, something else?
- readme: we could provide description on setting things up w/ `pyenv`... both `.python_version` as well as `pyenv virtualenv`

**Comment by hannahker** [comment]
> include a dummy .env file?

We have to be very careful about keeping that gitignored so I don't want to include a dummy file. Convention would be to document what it looks like in the README, which we can do once we have environment variables that we need.

> should we have the requirements-dev.txt file pre-built with black, isort, python-dotenv, jupyter-black, something else?

Yes good idea!

> readme: we could provide description on setting things up w/ pyenv... both .python_version as well as pyenv virtualenv

Good with me if that's our preferred Python environment management. Would need someone else to tackle that once since I'm a `venv` girl through and through.

**Comment by PaulineNimo** [comment]
One major question:
- Will this be the structure for all repos including those with the analysis in the qmds, notebooks and markdowns? 
Why I ask is because it may not exactly be smooth if the files are split between the folder depending on the language used. Just an opinion. 
My suggestion would be to have these two for any scripts such as codab loading etc and have one with the analysis at the top level with the exploration notebooks/markdowns with some form of flow or order.

**Comment by martinig94** [comment]
@hannahker Thanks a ton for the amazing work!


> > include a dummy .env file?
> 
> We have to be very careful about keeping that gitignored so I don't want to include a dummy file. Convention would be to document what it looks like in the README, which we can do once we have environment variables that we need.

Yes, also agree on not having a dummy .env file! 


> 
> > readme: we could provide description on setting things up w/ pyenv... both .python_version as well as pyenv virtualenv
> 
> Good with me if that's our preferred Python environment management. Would need someone else to tackle that once since I'm a `venv` girl through and through.

Same here, very used to venv but happy to switch in case

Question on my end:
Is the final idea to make packages out of this repo? If yes I am not sure if it's wise to keep multiple languages together. Thoughts? For what concerns R, would be Zack the main (and only?) developer and maintainer?

**Comment by zackarno** [comment]
- yeah good point on not including the `.env` - of course we don't want to push by accident
- re `.venv` happy either way, but at some point Tristan should show us how he uses `pyenv virtualenv` in his workflow because it's super nice and i started using it, but he has much more knowledge to describe all the pros-cons. This decision might also impact how we use jupyter notebooks and interact w/ kernels so it could be good to just do a bit more knowledge sharing on this before we come to a decision/if we do.
- Re R & python - my interpretation of the split here is not that it is a split we would follow in our own repos, this is just to make it clear which templates/code sit where in this repo. However, it is an option to merge them together as it would be pretty clear from the `.R` vs `.py` suffix anyways. Anyways happy for R not to be included here and we can make another one for R or work directly in a package (i.e [cumulus](https://github.com/OCHA-DAP/cumulus))
- I doubt we will ever turn this repo directly in to a package for many reasons, but also because it doesn't adhere to the package naming rules. So i don't think we have to worry about that aspect with regards to how the repo is structured/multi-language.

**Comment by hannahker** [comment]
Thanks all for the comments! 

- @PaulineNimo the structure here is specific to this repo and it's aims to be a store for general examples that we'll eventually turn into a package. I think the structure that you've suggested absolutely makes sense for other repos where we're doing analysis. 
- Agree it would be great to do some knowledge-sharing around Python env management. What about addressing that in the cookiecutter repo and maybe building up some guidance docs on Confluence? I don't think it's something we need to be prescriptive about across the team, but we should absolutely have some guidance there. 

> I doubt we will ever turn this repo directly in to a package for many reasons, but also because it doesn't adhere to the package naming rules. So i don't think we have to worry about that aspect with regards to how the repo is structured/multi-language.

Also agreed on this point! I'd imagined we would archive this repo once we're ready to create a package, and so just use this as a starting point.

**Comment by t-downing** [comment]
Sorry I am late to the game here!

In response:
- `.env`- agreed to not include, let's not risk it getting into the repo
- `venv` vs `pyenv virtualenv`- I used to use `venv` but then switched to `pyenv virtualenv` because it felt like a cleaner way to manage multiple python versions as well as virtual environments. I think it's fine if we use different setups, but I can add in specific `pyenv virtualenv` instructions in the `README.md` if we want.
- overall structure- yeah we wouldn't directly turn this repo into a package, but I think it still makes sense to structure the code as it would be in a package (within the top-level language directory), so then if we do create a package later it's much easier to just plunk this code in

Other points:
- do we want to use `ruff`? I've started using it and it's nice, but not tied to it

File: python/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Python Data Access Utilities
+
+This package provides utility functions for accessing our internal data infrastructure, including Azure Blob Storage and PostgreSQL databases.
+
+## Directory Structure
+
+```
+.
+├── requirements.txt        # Project dependencies
+├── setup.py                # Package configuration
```
Comment (t-downing) [inline python/README.md]: `setup.cfg` in the current structure

File: python/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+├── pyproject.toml          # Package configuration
+├── src/                    # Source code for utilities 
+└── examples/               # Usage examples
```
Comment (t-downing) [inline python/README.md:14]: Might also want to add `requirements-dev.txt` for completeness

File: python/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+1. Create and activate a virtual environment:
+```bash
+python -m venv venv
+source venv/bin/activate  # On Windows: venv\Scripts\activate
+```
+
+2. Install dependencies:
+```bash
+pip install -r requirements.txt
```
Comment (t-downing) [inline python/README.md:38]: Should this be `requirements-dev.txt`?

File: python/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+3. Install package in development mode:
+```bash
+pip install -e .
+```
+
+4. Set up the following environment variables in a `.env` file:
+```
+# TODO
+```
```
Comment (t-downing) [inline python/README.md:50]: We may also want instructions on creating a jupyter kernel- or should we just let people handle this on their own?

File: python/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+└── examples/               # Usage examples
+```
+
+## Developer Setup
+
+1. Create and activate a virtual environment:
+```bash
+python -m venv venv
+source venv/bin/activate  # On Windows: venv\Scripts\activate
+```
```
Comment (t-downing) [inline python/README.md:26]: The equivalent for `pyenv virtualenv` would be:

```bash
pyenv virtualenv <PYTHON_VERSION> virtualenv <VIRTUALENV_NAME>
pyenv local <VIRTUALENV_NAME>
pyenv activatate <VIRTUALENV_NAME>
```

For example:

```bash
pyenv virtualenv 3.11.4 virtualenv ds-my-repository
pyenv local ds-my-repository
pyenv activate ds-my-repository
```

Optionally you can add `eval "$(pyenv virtualenv-init -)"` to your `.zshrc` to automatically activate virtualenvs when you're in the directory.

File: python/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+2. Install dependencies:
+```bash
+pip install -r requirements.txt
+```
+
+3. Install package in development mode:
+```bash
+pip install -e .
+```
+
```
Comment (t-downing) [inline python/README.md:46]: We may also want to include `pre-commit install`?

File: python/README.md (hunk 5 above)
Comment (hannahker) [reply python/README.md:26]: Added!

File: python/README.md (hunk 4 above)
Comment (hannahker) [reply python/README.md:50]: For now, I think it's best to let people handle this on their own. It'll be useful to establish guidance here, but IMO we can do that in parallel as we start to build up the code in this repo.

**Comment by hannahker** [comment]
Thanks for the comments @t-downing! Have added in more references to `requirements-dev.txt` and `pre-commit`.  

I also like `ruff`! I'm still settling in to what the best setup for the `pyproject.toml` -- what do you use here?

File: python/README.md (hunk 5 above)
Comment (t-downing) [reply python/README.md:26]: Ah sorry just saw my typo in `pyenv activatate`...

**Comment by t-downing** [comment]
> I also like ruff! I'm still settling in to what the best setup for the pyproject.toml -- what do you use here?

I've used this, but still not sure what the best setup is either. So we can address later.

```
repos:
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.3.2
    hooks:
      # Run the linter
      - id: ruff
        args: [ --fix ]

      # Run the Ruff formatter
      - id: ruff-format
  - repo: https://github.com/psf/black
    rev: 24.10.0
    hooks:
      - id: black
```

**Pull request merged**